lib/chef/resource/lwrp_base.rb

```diff
@@ -35,32 +35,64 @@ class LWRPBase < Resource
 
       # Evaluates the LWRP resource file and instantiates a new Resource class.
       def self.build_from_file(cookbook_name, filename, run_context)
-        resource_class = nil
-        rname = filename_to_qualified_string(cookbook_name, filename)
-
-        class_name = convert_to_class_name(rname)
-        if Resource.const_defined?(class_name, false)
-          Chef::Log.info("#{class_name} light-weight resource is already initialized -- Skipping loading #{filename}!")
-          Chef::Log.debug("Overriding already defined LWRPs is not supported anymore starting with Chef 12.")
-          resource_class = Resource.const_get(class_name)
-        else
-          resource_class = Class.new(self)
+        resource_name = filename_to_qualified_string(cookbook_name, filename)
+
+        node = run_context ? run_context.node : Chef::Node.new
+        existing_class = Chef::ResourceResolver.new(node, resource_name).resolve
+
+        if existing_class
+          Chef::Log.info("Skipping LWRP resource #{filename} from cookbook #{cookbook_name}.  #{existing_class} already defined.")
+          Chef::Log.debug("Overriding resources with LWRPs is not supported anymore starting with Chef 12.")
+          return existing_class
+        end
 
-          Chef::Resource.const_set(class_name, resource_class)
-          resource_class.resource_name = rname
-          resource_class.run_context = run_context
-          resource_class.class_from_file(filename)
+        # We load the class first to give it a chance to set its own name
+        resource_class = Class.new(self)
```

Comment:
Er.  I *think* the code I wrote there does exactly what @sethvargo's did before, except now it uses the existence of a resource named `cookbookname_resourcename` as a marker, instead of looking for `Chef::Resource::CookbooknameResourcename`.  If anything, it seems like we will load the LWRP even less than we did before, because it's doing a real resolve that will take into account similarly-named resources that live in other namespaces.  Am I missing something?